### DISCUSSION TITLE
Multiple FK, references on the same table (Create messages table)

### DISCUSSION THREAD

**Discussion by david-shiko** [sqlalchemy/sqlalchemy#5901] [Usage Questions] [answered]
Please help me solve this error.
I just want to create a messages table with sender_id and recipient_id from users (foreign key constraint as only registered user may send massage)

```
class User(Base):
    __tablename__ = "users"
    # # # META # # #
    id = Column(Integer, primary_key=True, index=True)
    email = Column(String, unique=True, index=True, nullable=False)
    sent_messages = relationship("Message",  back_populates="users", cascade="all, delete")
    received_messages = relationship("Message",  back_populates="users", cascade="all, delete")


class Message(Base):
    __tablename__ = "messages"

    id = Column(Integer, primary_key=True)
    sender = Column(Integer, ForeignKey('users.id'), nullable=False)  # ForeignKey('users.id')
    recipient = Column(Integer, ForeignKey('users.id'), nullable=False)  # ForeignKey('users.id')
    data = Column(String, nullable=False)
    created_datetime = Column(DateTime, server_default=func.now())
    # # # RELATIONSHIPS # # #
    senders = relationship("User", foreign_keys=[sender], back_populates="sender")
    recipients = relationship("User", foreign_keys=[recipient], back_populates="recipient")
    
sqlalchemy.exc.AmbiguousForeignKeysError: Could not determine join condition between parent/child tables on relationship User.sent_messages - there are multiple foreign key paths linking the tables.  Specify the 'foreign_keys' argument, providing a list of those columns which should be counted as containing a foreign key reference to the parent table.
```

**Comment by gordthompson** [comment]
This is also being addressed on Stack Overflow [here](https://stackoverflow.com/q/65941555/2144390).

**Comment by david-shiko** [comment]
@gordthompson Can you please provide an example? I updated the question after your comment

**Comment by CaselIT** [comment, marked as answer]
Hi,

Looking at the error:

>sqlalchemy.exc.AmbiguousForeignKeysError: Could not determine join condition between parent/child tables on relationship User.sent_messages - there are multiple foreign key paths linking the tables.  Specify the 'foreign_keys' argument, providing a list of those columns which should be counted as containing a foreign key reference to the parent table.

the problem is that the relationship `sent_messages` (and `received_messages`) cannot determine the fk to `Message`. The reason is that there are 2 fk, `sender` and `recipient`. If you specify the `foreign_keys` in the relationship all should work.

Do you have some suggestion on how could the error message could be made clearer?

**Comment by david-shiko** [comment]
@CaselIT Hi!. But I specified relations and `FK` in both tables, where I should add one more `FK` or How I can specify which one should refer to which one?

I tried different ways but got an error every time

```
    # User table
    sent_messages = relationship("Message",  foreign_keys=['recipients'], back_populates="users", cascade="all, delete")
    received_messages = relationship("Message",  foreign_keys=['senders'], back_populates="users", cascade="all, delete")
```

or 

```
    # User table
    sent_messages = relationship("Message",  foreign_keys=[id], back_populates="users", cascade="all, delete")
    received_messages = relationship("Message",  foreign_keys=[id], back_populates="users", cascade="all, delete")
```

**Reply by CaselIT** [reply]
The problem is that the tables are linked by two fk, so the relationship does not know how what to use.

You should specify them in the user table relationships, using `foreign_keys=Message.recipients` or a lambda `foreign_keys=lambda: Message.recipients`. Same for `senders`.
There seems to be also another error with back_populates: it should specify the name of a relationship in the other model, not the table name. See also https://docs.sqlalchemy.org/en/13/orm/backref.html#relationships-backref and https://docs.sqlalchemy.org/en/13/orm/relationship_api.html?highlight=relationship#sqlalchemy.orm.relationship.params.back_populates

**Comment by gordthompson** [comment]
Answered on Stack Overflow [here](https://stackoverflow.com/a/65994756/2144390).

**Comment by Makistos** [comment]
I have the same issue so I'll hop in here if that's ok.

I have a table like this:

```
class Author(Base):
    __tablename__ = 'author'
    part_id = Column(Integer, ForeignKey('part.id'), nullable=False,
                     primary_key=True)
    person_id = Column(Integer, ForeignKey('person.id'), nullable=False,
                       primary_key=True)
    real_person_id = Column(Integer, ForeignKey('person.id'))
    parts = relationship("Part", backref=backref("part_assoc1"), viewonly=True)
    person = relationship("Person", foreign_keys=[person_id], backref=backref(
        "person1_assoc"), viewonly=True)
```

And I'm having an issue with this table:

```
class Part(Base):
    __tablename__ = 'part'
    id = Column(Integer, primary_key=True)
    authors = relationship('Person', secondary='author', foreign_keys=[Author.person_id],
                           uselist=True, viewonly=True)
```

Error I get is
`sqlalchemy.exc.NoForeignKeysError: Could not determine join condition between parent/child tables on relationship Part.authors - there are no foreign keys linking these tables via secondary table 'author'.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or specify 'primaryjoin' and 'secondaryjoin' expressions.`

So foreign_keys was not enough to solve this case. What else should I do?

**Reply by CaselIT** [reply]
The message is not very telling, you need to look for the source exception message. In your case it was
```
...
    raise exc.NoForeignKeysError(
sqlalchemy.exc.NoForeignKeysError: Can't find any foreign key relationships between 'part' and 'author'.

The above exception was the direct cause of the following exception:
...
```
adding it to the list of `foreign_keys` solved the error:
```py
class Part(Base):
    __tablename__ = 'part'
    id = Column(Integer, primary_key=True)
    authors = orm.relationship('Person', secondary='author', foreign_keys=[Author.part_id, Author.person_id],
                           uselist=True, viewonly=True, )
```

@zzzeek I'm not sure if it's possible to word the error better?

**Reply by Makistos** [reply]
Ok, this seems to have done the trick! Thank you! I never realised you actually need to give both keys in that relationship.

**Reply by CaselIT** [reply]
I was guessing based on the original error wording. The message can probably improved a bit
